polkadot/node/core/pvf/src/artifacts.rs

```diff
@@ -176,32 +211,107 @@ pub enum ArtifactState {
 
 /// A container of all known artifact ids and their states.
 pub struct Artifacts {
-	artifacts: HashMap<ArtifactId, ArtifactState>,
+	inner: HashMap<ArtifactId, ArtifactState>,
 }
 
 impl Artifacts {
-	/// Initialize a blank cache at the given path. This will clear everything present at the
-	/// given path, to be populated over time.
-	///
-	/// The recognized artifacts will be filled in the table and unrecognized will be removed.
-	pub async fn new(cache_path: &Path) -> Self {
-		// First delete the entire cache. This includes artifacts and any leftover worker dirs (see
-		// [`WorkerDir`]). Nodes are long-running so this should populate shortly.
-		let _ = tokio::fs::remove_dir_all(cache_path).await;
+	#[cfg(test)]
+	pub(crate) fn empty() -> Self {
+		Self { inner: HashMap::new() }
+	}
+
+	#[cfg(test)]
+	pub(crate) fn len(&self) -> usize {
+		self.inner.len()
+	}
+
+	/// Create an empty table and populate it with valid artifacts as [`ArtifactState::Prepared`],
+	/// if any. The existing caches will be checked by their file name to determine whether they are
+	/// valid, e.g., matching the current node version. The ones deemed invalid will be pruned.
+	pub async fn new_and_prune(cache_path: &Path) -> Self {
+		let mut artifacts = Self { inner: HashMap::new() };
+		artifacts.insert_and_prune(cache_path).await;
+		artifacts
+	}
+
+	async fn insert_and_prune(&mut self, cache_path: impl AsRef<Path>) {
+		fn is_stale(file_name: &str) -> bool {
+			!file_name.starts_with(ARTIFACT_PREFIX)
+		}
+
 		// Make sure that the cache path directory and all its parents are created.
+		let cache_path = cache_path.as_ref();
 		let _ = tokio::fs::create_dir_all(cache_path).await;
 
-		Self { artifacts: HashMap::new() }
-	}
+		match tokio::fs::read_dir(cache_path).await {
+			Ok(mut dir) => {
+				let mut prunes = vec![];
+
+				loop {
+					match dir.next_entry().await {
+						Ok(None) => break,
+						Ok(Some(entry)) => {
+							let file_name = entry.file_name();
+							if let Some(file_name) = file_name.to_str() {
+								let id = ArtifactId::from_file_name(file_name);
+								let file_path = cache_path.join(file_name);
+
+								if is_stale(file_name) || id.is_none() {
+									prunes.push(tokio::fs::remove_file(file_path.clone()));
+									gum::debug!(
+										target: LOG_TARGET,
```

Comment:
Nit: I have to quibble about 10 levels of indentation, although I'm doing that myself sometimes :slightly_smiling_face: 
The [style guide](https://github.com/paritytech/substrate/blob/master/docs/STYLE_GUIDE.md) says, "no more than 5 and definitely no more than 8". I'm in favor of relaxing that rule and increasing the allowed line length, but for now, it's there. But you don't have to follow it strictly.

Reply:
I agree. Maybe everything in the `Ok(Some(entry)) => {` block can be extracted into a helper. Would make it easier to read. :)

Reply:
This has to be refactored for sure :)

Reply:
Gonna punt on the help function. Having a function would complicate the typing a lot (`prunes` is a vec of futures). Compared to that, the indentation wouldn't be so bad.

Reply:
For the type you can just do `impl Future` without having to spell out the exact type. :)

I went ahead and refactored it in https://github.com/paritytech/polkadot-sdk/pull/1918/commits/4b1bb0ab6586d611da7201fb0ebb71c129feccb1.

Reply:
Okay, I get a typing error when I try to compile tests. Now I see what you mean.

TBH, I would keep the helper, but change it so that we call `.await` on `remove_file` immediately, instead of saving the future and calling `join_all` later. I'm sure it is not optimal, but this code should run very rarely and no one would notice it taking 0.1s longer, anyway. :P

Reply:
Ok, I'll keep on working with your code. But I'll probably keep the `new` function small, only handling construction as it was before :)

Reply:
Deep indentation is often bad because it can be a trait of stuff made unnecessarily harder to reason about. If that's the case, it should be uncluttered for sure, e.g., by returning early, which is often an improvement. The guideline is great for that purpose.

But sometimes, it's just what it takes, like in this case. The logic is simple: register if valid, otherwise prune. The rest is kind of collateral. And even with that, the code isn’t that hard to reason about. We’re not really simplifying stuff this way, just relocating. Idk, is it worth it to make that compromise, for aesthetic reasons?

Reply:
Good point. Personally, comparing before and after, I do find it easier to read now. And to me it makes sense to separate out handling of individual entries into a helper function. `insert_and_prune` works on a whole directory, while `insert_or_prune` works on individual entries. I can see it being subjective, but since we have a style guideline, I'd say let's just follow it.

Reply:
I like how you organize helper functions as subfunctions, we should do that more often!